### PR TITLE
Various fixes for Tribler 7

### LINKED ISSUE #3038: Request.finish called on a request after its connection was lost
- [X] I have read CONTRIBUTING.rst

- [X] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
$ git status
On branch next
Your branch is up-to-date with 'origin/next'.
nothing to commit, working directory clean

##### Operating system and version: #####
Ubuntu 16.04

Checking my log files for errors after I received a couple of them the last couple of days. This error I encountered once (yesterday I think, not sure as tribler.log does not log any timestamps). I am runing tribler next in a docker container which I connect to through vnc/novnc.

```
Unhandled error in Deferred:


Traceback (most recent call last):
  File "/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py", line 487, in _task_process_alerts
    self.process_alert(alert)
  File "/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py", line 336, in process_alert
    self.torrents[info_hash][0].deferred_removed.callback(None)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 393, in callback
    self._startRunCallbacks(result)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 588, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/opt/tribler/Tribler/Core/Modules/restapi/downloads_endpoint.py", line 336, in _on_remove_failure
    request.finish()
  File "/usr/lib/python2.7/dist-packages/twisted/web/server.py", line 224, in finish
    return http.Request.finish(self)
  File "/usr/lib/python2.7/dist-packages/twisted/web/http.py", line 865, in finish
    "Request.finish called on a request after its connection was lost; "
exceptions.RuntimeError: Request.finish called on a request after its connection was lost; use Request.notifyFinish to keep track of this.
ERROR:root:Traceback (most recent call last):
  File "/opt/tribler/TriblerGUI/event_request_manager.py", line 99, in on_read_data
    raise RuntimeError(json_dict["event"]["text"])
RuntimeError: [Failure instance: Traceback: <type 'exceptions.RuntimeError'>: Request.finish called on a request after its connection was lost; use Request.notifyFinish to keep track of this.
/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py:487:_task_process_alerts
/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py:336:process_alert
/usr/lib/python2.7/dist-packages/twisted/internet/defer.py:393:callback
/usr/lib/python2.7/dist-packages/twisted/internet/defer.py:501:_startRunCallbacks
--- <exception caught here> ---
/usr/lib/python2.7/dist-packages/twisted/internet/defer.py:588:_runCallbacks
/opt/tribler/Tribler/Core/Modules/restapi/downloads_endpoint.py:336:_on_remove_failure
/usr/lib/python2.7/dist-packages/twisted/web/server.py:224:finish
/usr/lib/python2.7/dist-packages/twisted/web/http.py:865:finish
```

### LINKED ISSUE #3039: AttributeError: 'NoneType' object has no attribute 'get_peer_info'
- [X] I have read CONTRIBUTING.rst

- [X] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
$ git status
On branch next
Your branch is up-to-date with 'origin/next'.
nothing to commit, working directory clean

##### Operating system and version: #####
Ubuntu 16.04

Checking my log files for errors after I received a couple of them the last couple of days. This error I encountered multiple times. I am runing tribler next in a docker container which I connect to through vnc/novnc. I was looking at [this](http://urlgone.com/67a007/) topic when this error happened repeatedly.

```
Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/tribler/TriblerGUI/core_manager.py", line 63, in start_tribler_core
    reactor.run()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1194, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1203, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/opt/tribler/Tribler/dispersy/requestcache.py", line 213, in _on_timeout
    cache.on_timeout()
  File "/opt/tribler/Tribler/community/tunnel/tunnel_community.py", line 61, in on_timeout
    self.community.remove_circuit(self.number, reason)
  File "/opt/tribler/Tribler/community/tunnel/hidden_community.py", line 203, in remove_circuit
    super(HiddenTunnelCommunity, self).remove_circuit(circuit_id, additional_info, destroy)
  File "/opt/tribler/Tribler/community/tunnel/tunnel_community.py", line 624, in remove_circuit
    for d, s in ltmgr.torrents.values() if s == ltmgr.get_session(d.get_hops())}
  File "/opt/tribler/Tribler/community/tunnel/tunnel_community.py", line 624, in <dictcomp>
    for d, s in ltmgr.torrents.values() if s == ltmgr.get_session(d.get_hops())}
exceptions.AttributeError: 'NoneType' object has no attribute 'get_peer_info'
```

### LINKED ISSUE #3040: KeyError: Unhandled error in Deferred:
- [X] I have read CONTRIBUTING.rst

- [X] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
$ git status
On branch next
Your branch is up-to-date with 'origin/next'.
nothing to commit, working directory clean

##### Operating system and version: #####
Ubuntu 16.04

Checking my log files for errors after I received a couple of them the last couple of days (sorry for spamming these issues). This error I encountered a couple of times as well. I am runing tribler next in a docker container which I connect to through vnc/novnc.

```
Unhandled error in Deferred:


Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1194, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1203, in mainLoop
    self.runUntilCurrent()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/task.py", line 239, in __call__
    d = defer.maybeDeferred(self.f, *self.a, **self.kw)
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py", line 487, in _task_process_alerts
    self.process_alert(alert)
  File "/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py", line 336, in process_alert
    self.torrents[info_hash][0].deferred_removed.callback(None)
exceptions.KeyError: '26ac4cda53aabccc6377343f4c84c9097c7d15eb'
ERROR:root:Traceback (most recent call last):
  File "/opt/tribler/TriblerGUI/event_request_manager.py", line 99, in on_read_data
    raise RuntimeError(json_dict["event"]["text"])
RuntimeError: [Failure instance: Traceback: <type 'exceptions.KeyError'>: '26ac4cda53aabccc6377343f4c84c9097c7d15eb'
/usr/lib/python2.7/dist-packages/twisted/internet/base.py:1194:run
/usr/lib/python2.7/dist-packages/twisted/internet/base.py:1203:mainLoop
/usr/lib/python2.7/dist-packages/twisted/internet/base.py:825:runUntilCurrent
/usr/lib/python2.7/dist-packages/twisted/internet/task.py:239:__call__
--- <exception caught here> ---
/usr/lib/python2.7/dist-packages/twisted/internet/defer.py:150:maybeDeferred
/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py:487:_task_process_alerts
/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py:336:process_alert
]
```

### LINKED ISSUE #3041: IOError: [Errno 2] No usable temporary directory
- [X] I have read CONTRIBUTING.rst

- [X] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
$ git status
On branch next
Your branch is up-to-date with 'origin/next'.
nothing to commit, working directory clean

##### Operating system and version: #####
Ubuntu 16.04

Checking my log files for errors after I received a couple of them the last couple of days (sorry for spamming these issues).  This error I received when I was setting up my docker container and was easy to fix, but could maybe be nicer handled as I didnot get some sort of popup notification about it and tribler just kept crashing/shutting off.

```
See the manual page for dbus-uuidgen to correct this issue.
ERROR:root:Traceback (most recent call last):
  File "run_tribler.py", line 21, in <module>
    window = TriblerWindow()
  File "/opt/tribler/TriblerGUI/tribler_window.py", line 106, in __init__
    uic.loadUi(get_ui_file_path('mainwindow.ui'), self)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/__init__.py", line 222, in loadUi
    return DynamicUILoader(package).loadUi(uifile, baseinstance, resource_suffix)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/Loader/loader.py", line 72, in loadUi
    return self.parse(filename, resource_suffix, basedir)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 980, in parse
    actor(elem)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 787, in createUserInterface
    self.traverseWidgetTree(elem)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 765, in traverseWidgetTree
    handler(self, child)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 249, in createWidget
    self.traverseWidgetTree(elem)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 765, in traverseWidgetTree
    handler(self, child)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 462, in createLayout
    self.traverseWidgetTree(elem)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 765, in traverseWidgetTree
    handler(self, child)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 505, in handleItem
    self.traverseWidgetTree(elem)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 765, in traverseWidgetTree
    handler(self, child)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 462, in createLayout
    self.traverseWidgetTree(elem)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 765, in traverseWidgetTree
    handler(self, child)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 505, in handleItem
    self.traverseWidgetTree(elem)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 765, in traverseWidgetTree
    handler(self, child)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 249, in createWidget
    self.traverseWidgetTree(elem)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 765, in traverseWidgetTree
    handler(self, child)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 240, in createWidget
    self.stack.push(self.setupObject(widget_class, parent, elem))
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/uiparser.py", line 204, in setupObject
    obj = self.factory.createQObject(clsname, name, args, is_attribute)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/objcreator.py", line 106, in createQObject
    factory = self.findQObjectType(classname)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/objcreator.py", line 131, in findQObjectType
    w = module.search(classname)
  File "/usr/lib/python2.7/dist-packages/PyQt5/uic/Loader/qobjectcreator.py", line 115, in search
    module = __import__(mname, {}, {}, (cls,))
  File "/opt/tribler/TriblerGUI/widgets/trustpage.py", line 3, in <module>
    import matplotlib
  File "/usr/lib/python2.7/dist-packages/matplotlib/__init__.py", line 1133, in <module>
    rcParams = rc_params()
  File "/usr/lib/python2.7/dist-packages/matplotlib/__init__.py", line 967, in rc_params
    fname = matplotlib_fname()
  File "/usr/lib/python2.7/dist-packages/matplotlib/__init__.py", line 796, in matplotlib_fname
    configdir = _get_configdir()
  File "/usr/lib/python2.7/dist-packages/matplotlib/__init__.py", line 649, in _get_configdir
    return _get_config_or_cache_dir(_get_xdg_config_dir())
  File "/usr/lib/python2.7/dist-packages/matplotlib/__init__.py", line 626, in _get_config_or_cache_dir
    return _create_tmp_config_dir()
  File "/usr/lib/python2.7/dist-packages/matplotlib/__init__.py", line 550, in _create_tmp_config_dir
    tempdir = tempfile.gettempdir()
  File "/usr/lib/python2.7/tempfile.py", line 275, in gettempdir
    tempdir = _get_default_tempdir()
  File "/usr/lib/python2.7/tempfile.py", line 217, in _get_default_tempdir
    ("No usable temporary directory found in %s" % dirlist))
IOError: [Errno 2] No usable temporary directory found in ['/tmp', '/var/tmp', '/usr/tmp', '/opt/tribler']
```

### LINKED ISSUE #3042: TypeError: Non-base32 digit found
- [X] I have read CONTRIBUTING.rst

- [X] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
$ git status
On branch next
Your branch is up-to-date with 'origin/next'.
nothing to commit, working directory clean

##### Operating system and version: #####
Ubuntu 16.04

Another one I just encoutered, happened with [this](http://mgnet.me/cwt09E6) torrent. Btw, the error message is directly displayed in the 'Download torrent' popup where normally the files within the torrent are displayed. The error message is very long and is not broken on to a new line, so the popup is veeeerrrryyy width and actually goes off screen.

```
ERROR:RESTRequest:[Failure instance: Traceback: <type 'exceptions.TypeError'>: Non-base32 digit found
/usr/lib/python2.7/dist-packages/twisted/protocols/basic.py:571:dataReceived
/usr/lib/python2.7/dist-packages/twisted/web/http.py:1657:lineReceived
/usr/lib/python2.7/dist-packages/twisted/web/http.py:1736:allContentReceived
/usr/lib/python2.7/dist-packages/twisted/web/http.py:762:requestReceived
--- <exception caught here> ---
/usr/lib/python2.7/dist-packages/twisted/web/server.py:183:process
/usr/lib/python2.7/dist-packages/twisted/web/server.py:234:render
/usr/lib/python2.7/dist-packages/twisted/web/resource.py:250:render
/opt/tribler/Tribler/Core/Modules/restapi/torrentinfo_endpoint.py:102:render_GET
/opt/tribler/Tribler/Core/Utilities/utilities.py:249:parse_magnetlink
/usr/lib/python2.7/base64.py:229:b32decode
]
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/twisted/web/server.py", line 183, in process
    self.render(resrc)
  File "/usr/lib/python2.7/dist-packages/twisted/web/server.py", line 234, in render
    body = resrc.render(self)
  File "/usr/lib/python2.7/dist-packages/twisted/web/resource.py", line 250, in render
    return m(request)
  File "/opt/tribler/Tribler/Core/Modules/restapi/torrentinfo_endpoint.py", line 102, in render_GET
    self.infohash = parse_magnetlink(uri)[1]
  File "/opt/tribler/Tribler/Core/Utilities/utilities.py", line 249, in parse_magnetlink
    xt = b32decode(encoded_infohash)
  File "/usr/lib/python2.7/base64.py", line 229, in b32decode
    raise TypeError('Non-base32 digit found')
TypeError: Non-base32 digit found
```

### LINKED ISSUE #3044: RuntimeError: torrent already exists in session
- [X] I have read CONTRIBUTING.rst

- [X] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
$ git status
On branch next
Your branch is up-to-date with 'origin/next'.
nothing to commit, working directory clean

##### Operating system and version: #####
Ubuntu 16.04

Just happened while testing #3043, tribler crashed with the following message. I had one torrent that was still 'building circuits' (eg didnot list a name in the donwload overview). After disabling anonimity for all torrents tribler crashed and after a restart that torrent was gone from the overview.

```
Traceback (most recent call last):
  File "/opt/tribler/TriblerGUI/event_request_manager.py", line 99, in on_read_data
    raise RuntimeError(json_dict["event"]["text"])
RuntimeError: Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/tribler/TriblerGUI/core_manager.py", line 63, in start_tribler_core
    reactor.run()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1194, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1203, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/opt/tribler/Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py", line 240, in schedule_create_engine
    self.pstate_for_restart, share_mode=share_mode)
  File "/opt/tribler/Tribler/Core/Libtorrent/LibtorrentDownloadImpl.py", line 340, in network_create_engine_wrapper
    self.handle = self.ltmgr.add_torrent(self, atp)
  File "/opt/tribler/Tribler/Core/Libtorrent/LibtorrentMgr.py", line 278, in add_torrent
    torrent_handle = ltsession.add_torrent(encode_atp(atp))
exceptions.RuntimeError: torrent already exists in session
```

### LINKED ISSUE #3046: Missing torrent after crash
- [X] I have read CONTRIBUTING.rst

- [X] I have tried with the latest pre-release version and I still can reproduce the issue.

##### Tribler version/branch+revision: #####
$ git status
On branch next
Your branch is up-to-date with 'origin/next'.
nothing to commit, working directory clean

##### Operating system and version: #####
Ubuntu 16.04

This morning I was cleaning up the finished torrents, I had only one torrent still downloading at ~5KB/s. After removing two torrents from the GUI, tribler crashed as mentioned in #3039 But after the restart the torrent that was still downloading is gone. Files still exists in the download folder but the torrent is not listed anymore.

Looking at the logs the only thing that is strange is that currently the log is filled with a lot of these messages. These messages are not showing before the crash, but after the crash they did.
```
WARNING:SearchCommunity:unable to find conversion to decode 0001fb04df93369587ec8fd9b74559186fa356cffda8f561f1730ce2b03f1ed74582504d684a078d0ece4700000000000000025ed0001e1e4f83b41b9b191783b41b9b1917c0a8010c1e4f4e2d7ace1e4f00611503623b515f7f125bc9748c39a3b2678d6c3092b7f6007831d126f308c63f242e549784a84ea16dc55932 in [<DefaultConversion 0000 [dispersy-dynamic-settings, dispersy-missing-identity, dispersy-missing-sequence, dispersy-missing-message, dispersy-puncture-request, dispersy-undo-other, dispersy-signature-response, dispersy-destroy-community, dispersy-introduction-response, dispersy-identity, dispersy-signature-request, dispersy-authorize, dispersy-puncture, dispersy-missing-proof, dispersy-revoke, dispersy-undo-own, dispersy-introduction-request]>, <SearchConversion 0002 [dispersy-missing-message, torrent-collect-request, dispersy-signature-response, search-response, dispersy-missing-identity, torrent-collect-response, dispersy-puncture-request, dispersy-undo-other, dispersy-dynamic-settings, dispersy-undo-own, dispersy-introduction-request, dispersy-missing-sequence, dispersy-destroy-community, dispersy-introduction-response, dispersy-identity, dispersy-signature-request, dispersy-revoke, dispersy-puncture, torrent-request, search-request, dispersy-authorize, dispersy-missing-proof, torrent]>]
```

### PR DESCRIPTION
**Pull request by qstokkink**
Fixes #3038, fixes #3039, fixes #3040, fixes #3041, fixes #3042, fixes #3044, fixes #3046.

### CODE REVIEW

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

File: Tribler/Core/Libtorrent/LibtorrentMgr.py [hunk 1]
```diff
@@ -275,6 +275,13 @@ def add_torrent(self, torrentdl, atp):
                 if request_handle:
                     ltsession.remove_torrent(request_handle, 0)
 
+            # Check if we added this torrent before
+            known = ltsession.find_torrent(lt.sha1_hash(infohash))
```
Comment (devos50) [inline Tribler/Core/Libtorrent/LibtorrentMgr.py]: What is `known` is `None`? Or is it impossible for this to happen?

Comment (qstokkink) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: According to the documentation it always returns a handle, but `is_valid()` will return `False` if the torrent could not be found: http://www.rasterbar.com/products/libtorrent/manual.html#find-torrent-get-torrents , https://github.com/arvidn/libtorrent/blob/ae0baff5a399e75b08d1aa2ac27f84b641c6426b/include/libtorrent/session_handle.hpp#L185 . I tried finding the cpp line as well, but the libtorrent source is a bit too cryptic for me to casually interpret.

Comment (qstokkink) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: **tl;dr** Yes it is impossible.

Comment (devos50) [reply Tribler/Core/Libtorrent/LibtorrentMgr.py]: Okay great! I guess this one can be merged then 👍

**Pull request merged**